hudi-common/src/test/java/org/apache/hudi/common/util/TestHoodieAvroUtils.java

```diff
@@ -57,4 +60,16 @@ public void testPropsPresent() {
     }
     Assert.assertTrue("column pii_col doesn't show up", piiPresent);
   }
+
+  @Test
+  public void testDefaultValue() {
+    GenericRecord rec = new GenericData.Record(new Schema.Parser().parse(EXAMPLE_SCHEMA));
+    rec.put("_row_key", "key1");
+    rec.put("non_pii_col", "val1");
+    rec.put("pii_col", "val2");
+    rec.put("timestamp", 3.5);
```

Comment:
So the issue seems to be that in the original record created in this way, the default values shows up as `null`. Even though you have specified `default: dummy_val` it still is showing up as `null` in the original record.

Do you know why that is the case ? When we have specified the default value, why doesn't Avro put it in the record when the field is missing ?

I tried using the builder, but that expects default values to be specified for each and every field else throws an excpetion:
```
GenericRecord rec = new GenericRecordBuilder(new Schema.Parser().parse(EXAMPLE_SCHEMA)).build();
```
Do you have more research points around why this is the case with Avro ?

Reply:
No its not that the original record has default values as null. Its just while getting the values from the record, default values are not considered. Please have a look at this function - 

@Override public Object get(String key) {
      Field field = schema.getField(key);
      if (field == null) return null;
      return values[field.pos()];
    }

Ideally the above function should get field.defaultVal() in case values[field.pos()] is null, but that is not the case.

Reply:
Can you help me understand how you are running into this issue with default values ? 

Based on my understanding, conversion to avro is internal to Hudi and a custom avro schema (with default values) is not something that user can themselves pass. And how `spark-avro` converts `struct schema to avro` there is no special handling there from `default value` perspective. So I guess I am not sure whether this is an issue in the first place.

Reply:
> conversion to avro is internal to Hudi and a custom avro schema (with default values) is not something that user can themselves pass

I did not understand this. As a user I can always specify the schema that is to be used either via FileBasedSchemaProvider or using schema registry. 

Let me give you an example. Suppose there is some table with schema S1 and you have published some records (R1 and R2) with this schema into kafka. Next you evolve the schema (it now becomes S2) and a new nullable field is added as below -> 

{"name": "col1", "type":["string", "null"], "default": "dummy"}

you again publish some records (R3 and R4) with S2 and now start consuming with delta streamer. So your kafka topic is having records with both the schemas and delta streamer is using S2 as target schema. Now while writing to parquet, I want R1 and R2 to be written with this default value "dummy" for field "col1", which is a pretty common case. Generally users prefer to have some default value for newly added fields rather than having written them as null. How do you achieve this without this PR? 

Open to hearing your thoughts on this.

Reply:
My bad I was thinking only from `DataSource's HoodieSparkSqlWriter` writer point of view, where the schema is determined automatically from the `DataFrame` and converted to avro schema. Missed that `DeltaStreamer` uses the `schema provider` which the users can pass it directly to the `HoodieWriteClient`. Thanks for details !

I have a question for the schema evolution example you provided. The `rewriteRecord()` you are testing here uses the schema from the old record, and re-writes by setting only the fields found in the old schema. So if you rewrite R1 and R2 record, there schema will not have the new `col1` field right ? Hence, your code of populating default values will not get executed because `col1` is not present in the old schema fields.

It seems this test case works because you are not evolving the schema here. Your old and new record both have the same schema. But if your old record schema is different I think you will run into the same issue. Am I missing something here ?

Reply:
My bad. You raised a valid point there. Have made few changes and added more test cases to cover schema evolution scenario as well. Please take a pass @umehrot2